#**Cricket - Test Matches Analysis**
Cricket is one of the most popular spectator sports in the world with viewership in billions spread across the globe. Test cricket format is the toughest of all in cricket format due to longer time spend, unlimited bow and others. By Test Mathces analysis, the quality of a Player can be easily understandable. This is because if you want to survive in the longest format you will be needing a lot more qualities than you need in the shorter formats. For this format, cricketers need to be patient, aggressive, good readers of the game and a good thinker. Only a few people in the world are this qualified and talented.
In this notebook, I will do a comprehensive analysis of the Test Cricket Matches by comparing over different categories. I'll look for insights in the data to devise strategies to findout world most qualified and talented playes in Test Cricket.

### **By- Md. Azim Khan**

                                                                

In [482]:
# importing data analysis libraries
import pandas as pd
import numpy as np

In [483]:
# read dataset from csv file
df = pd.read_csv('/content/drive/MyDrive/wickets.csv')

In [484]:
# display the first few rows of the data
display(df.head(10))


Player       Span   Mat  Inns  ...  Econ    SR   5  10
0  M Muralitharan (ICC/SL)  1992-2010   133   230  ...  2.47  55.0  67  22
1           SK Warne (AUS)  1992-2007   145   273  ...  2.65  57.4  37  10
2        JM Anderson (ENG)  2003-2021  164*   304  ...  2.83  56.3  30   3
3         A Kumble (INDIA)  1990-2008   132   236  ...  2.69  65.9  35   8
4         GD McGrath (AUS)  1993-2007   124   243  ...  2.49  51.9  29   3
5          SCJ Broad (ENG)  2007-2021   149   274  ...  2.93  56.9  18   3
6            CA Walsh (WI)  1984-2001   132   242  ...  2.53  57.8  22   3
7            DW Steyn (SA)  2004-2019    93   171  ...  3.24  42.3  26   5
8      N Kapil Dev (INDIA)  1978-1994   131   227  ...  2.78  63.9  23   2
9        HMRKB Herath (SL)  1999-2018    93   170  ...  2.80  60.0  34   9

[10 rows x 14 columns]

In [485]:
# display the all features or variables or all columns name 
df.columns


Index(['Player', 'Span', 'Mat', 'Inns', 'Balls', 'Runs', 'Wkts', 'BBI', 'BBM',
       'Ave', 'Econ', 'SR', '5', '10'],
      dtype='object')

# **Explanation of Feature variables**


1. Player = Name of a bowler name
2. Span =  Playing span (difference bewteen Starting year Ending year)
3. Mat =  Total number of Match played by a bowler
4. Inns =  Total number of Innings bowled by a bowler (One Match = 2 innings_)
5. Balls = Total number of Balls bowled by a bowler
6. Runs = Total runs have been conceded by a bowler
7. Wkts = Total wicket has been taken by a bowler in his career
8. BBI = BBI stands for Best Bowling in Innings  for a bowler
9. BBM = BBM stands for Best Bowling in Match and gives the combined score over 2 or more innings in one match.
10. Ave = bowling average ( like average 22.72 means one bowler take one wicket after giving 22.72 runs)
11. Econ = Economy rate means a bowler conceded run by one over (Econ 2.47 means one bowler conceded 2.47 run per over)

12. SR = balls per wicket (SR 55 means one bowler take one wicket after 55 bowled)
13. 5 = The total number of 5 wickets has taken in an Innings by a bowler
14. 10 =  The total number of 10 wickets has taken in a Match by a bowler




In [486]:
#data types and missing value check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     object 
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     79 non-null     object 
 8   BBM     79 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(6), object(5)
memory usage: 8.8+ KB


# **Observations:**

* Number of observations (rows) = 79
* Number of columns = 14
* The dataset has no missing values
* memory usage 8.8 kilobyte

In [487]:
# checking data statistics
df.describe()

Inns         Balls          Runs  ...         SR          5         10
count   79.000000     79.000000     79.000000  ...  79.000000  79.000000  79.000000
mean   144.911392  18638.354430   8599.354430  ...  59.193671  16.354430   2.797468
std     51.180222   7199.256972   3085.168807  ...   9.350132   9.642372   3.235935
min     67.000000   8785.000000   4846.000000  ...  41.200000   3.000000   0.000000
25%    110.000000  13583.000000   6456.500000  ...  53.300000   9.500000   1.000000
50%    129.000000  16498.000000   7742.000000  ...  57.400000  14.000000   2.000000
75%    169.000000  21742.500000   9756.000000  ...  63.950000  20.500000   3.500000
max    304.000000  44039.000000  18355.000000  ...  91.900000  67.000000  22.000000

[8 rows x 9 columns]

# **Data statistics :**

* The average wicket taken by a bowler in the dataset is 317

* The one wicket has taken after the Highest 91 balls by a bowler.
* The average runs has been conceded by a bowler is 8599 

* The most expensive player is  3.46 (from the Economy rate)

* the highest number of innigs played by a bowler is 304


# **Some Questions and Answer**

### **Q1.   How to remove the BBI and BBM columns from dataset?**




In [488]:
# use drop method
df1 = df.drop(['BBI','BBM'], axis = 1)
df1.head(5)

Player       Span   Mat  Inns  ...  Econ    SR   5  10
0  M Muralitharan (ICC/SL)  1992-2010   133   230  ...  2.47  55.0  67  22
1           SK Warne (AUS)  1992-2007   145   273  ...  2.65  57.4  37  10
2        JM Anderson (ENG)  2003-2021  164*   304  ...  2.83  56.3  30   3
3         A Kumble (INDIA)  1990-2008   132   236  ...  2.69  65.9  35   8
4         GD McGrath (AUS)  1993-2007   124   243  ...  2.49  51.9  29   3

[5 rows x 12 columns]

**Remark:using drop method, it is possible to remove columns**  



### **Q2.   How many players played for ICC?**




In [489]:
# here working with df1 dataset not orginal dataset
# splitting the 'Player' column to get the information about 'ICC' and 'Country
df_player = df1['Player'].str.split("(", expand=True)
# join 'Country' with the main dataset
df1 = pd.concat([df1, df_player], axis=1)
# dropping the 'Player' columns, after spliting , it is no need
df1 = df1.drop('Player', axis=1)

# renaming the column names
df1 = df1.rename(columns={0: 'Player',
                        1: 'Country',
                        '5' : '5_wkts',
                        '10' :  '10_wkts'      })
display(df1.head(5))  

Span   Mat  Inns  Balls  ...  5_wkts  10_wkts           Player  Country
0  1992-2010   133   230  44039  ...      67       22  M Muralitharan   ICC/SL)
1  1992-2007   145   273  40705  ...      37       10        SK Warne      AUS)
2  2003-2021  164*   304  35079  ...      30        3     JM Anderson      ENG)
3  1990-2008   132   236  40850  ...      35        8        A Kumble    INDIA)
4  1993-2007   124   243  29248  ...      29        3      GD McGrath      AUS)

[5 rows x 13 columns]

In [490]:
# remove the ")" from the 'Country' column
df1['Country'] = df1['Country'].str.replace(")", "")

# rearrange the columns
new_col_sequence = ['Player', 'Span', 'Mat', 'Inns', 'Balls', 'Runs', 'Wkts','Ave','Econ','SR','5_wkts', '10_wkts','Country']
df1 = df1[new_col_sequence]
display(df1.head(5))

Player       Span   Mat  Inns  ...    SR  5_wkts  10_wkts  Country
0  M Muralitharan   1992-2010   133   230  ...  55.0      67       22   ICC/SL
1        SK Warne   1992-2007   145   273  ...  57.4      37       10      AUS
2     JM Anderson   2003-2021  164*   304  ...  56.3      30        3      ENG
3        A Kumble   1990-2008   132   236  ...  65.9      35        8    INDIA
4      GD McGrath   1993-2007   124   243  ...  51.9      29        3      AUS

[5 rows x 13 columns]

In [491]:
# Function create to check the icc
def icc_check(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"

In [492]:
# call the function 
df1['played_for_ICC'] = df1['Country'].apply(icc_check)

display(df1.head(5))

Player       Span   Mat  ...  10_wkts  Country  played_for_ICC
0  M Muralitharan   1992-2010   133  ...       22   ICC/SL             Yes
1        SK Warne   1992-2007   145  ...       10      AUS              No
2     JM Anderson   2003-2021  164*  ...        3      ENG              No
3        A Kumble   1990-2008   132  ...        8    INDIA              No
4      GD McGrath   1993-2007   124  ...        3      AUS              No

[5 rows x 14 columns]

In [493]:
# value_counts() method is used to check the number of unique value presents in the dataset
df1['played_for_ICC'].value_counts()

No     74
Yes     5
Name: played_for_ICC, dtype: int64

**Remark:** **only 5 playeres are played for the ICC**

### **Q3. How many different countries are present in this dataset?**

In [494]:
# First need to replace ICC from Country columns, as ICC is not a valid country name
df1['Country'] = df1['Country'].str.replace("ICC/", "")
df1['Country'] = df1['Country'].str.replace("/ICC", "")

display(df1.head())

Player       Span   Mat  ...  10_wkts  Country  played_for_ICC
0  M Muralitharan   1992-2010   133  ...       22       SL             Yes
1        SK Warne   1992-2007   145  ...       10      AUS              No
2     JM Anderson   2003-2021  164*  ...        3      ENG              No
3        A Kumble   1990-2008   132  ...        8    INDIA              No
4      GD McGrath   1993-2007   124  ...        3      AUS              No

[5 rows x 14 columns]

In [495]:
#again value_counts() method is used to know the unique values
df1['Country'].value_counts()

AUS      18
ENG      15
INDIA    10
WI        9
SA        8
NZ        7
PAK       7
SL        3
ZIM       1
BDESH     1
Name: Country, dtype: int64

**There are 10 different countries presents in the Dataset**

### **Q4.Which player(s) had played for the longest period of time?**

In [496]:
# splitting the 'Span' column based on the "-"
df_span = df1['Span'].str.split("-", expand=True)

# conccatinating the new dataframe with the main dataframe
df1 = pd.concat([df1, df_span], axis=1)

# renaming the newly created column names
df1 = df1.rename(columns={0: "start_year",
                        1: "end_year"})

# removing the "Span" column
df1 = df1.drop("Span", axis=1)

display(df1.head(5))

Player   Mat  Inns  ...  played_for_ICC  start_year  end_year
0  M Muralitharan    133   230  ...             Yes        1992      2010
1        SK Warne    145   273  ...              No        1992      2007
2     JM Anderson   164*   304  ...              No        2003      2021
3        A Kumble    132   236  ...              No        1990      2008
4      GD McGrath    124   243  ...              No        1993      2007

[5 rows x 15 columns]

In [497]:
# to check the datatypes for finding the span
df1.dtypes

Player             object
Mat                object
Inns                int64
Balls               int64
Runs                int64
Wkts                int64
Ave               float64
Econ              float64
SR                float64
5_wkts              int64
10_wkts             int64
Country            object
played_for_ICC     object
start_year         object
end_year           object
dtype: object

**Remark**: start_year and end_year both are string datatypes, it's not possible to find out the difference from this datatypes and need to convert

In [498]:
# convert the object data type into int using astype method 
df1['start_year'] = df1['start_year'].astype('int') 
df1['end_year'] = df1['end_year'].astype('int')

df1.dtypes

Player             object
Mat                object
Inns                int64
Balls               int64
Runs                int64
Wkts                int64
Ave               float64
Econ              float64
SR                float64
5_wkts              int64
10_wkts             int64
Country            object
played_for_ICC     object
start_year          int64
end_year            int64
dtype: object

**Remark:** now start_year and end_year both are int datatypes

In [499]:
# Findout the total year played by a bowler from two columns 
df1['years_played'] = df1['end_year'] - df1['start_year']
# after finding years_played, then no need to keep two column in dataset
df1 = df1.drop(['start_year', "end_year"], axis=1)



In [500]:
# sort the years_played column to findout which players had played for the longest period of time.
df1.sort_values('years_played', ascending = False).head(5)

Player   Mat  Inns  ...  Country  played_for_ICC  years_played
21      Imran Khan     88   142  ...      PAK              No            21
55       GS Sobers     93   159  ...       WI              No            20
9     HMRKB Herath     93   170  ...       SL              No            19
0   M Muralitharan    133   230  ...       SL             Yes            18
2      JM Anderson   164*   304  ...      ENG              No            18

[5 rows x 14 columns]

**Imran Khan, Pakistani player, had played for the longest period of time and it is 21 years.**

### **Q5. Which player(s) had played for the shortest period of time?**

In [501]:
# sort the years_played column keeping ascending true to findout which player had played for the shortest period of time.
df1.sort_values('years_played', ascending = True).head(5)

Player  Mat  Inns  ...  Country  played_for_ICC  years_played
44      GP Swann    60   109  ...      ENG              No             5
71      K Rabada    47    86  ...       SA              No             6
54    Yasir Shah   46*    84  ...      PAK              No             7
61   SJ Harmison    63   115  ...      ENG             Yes             7
72  JR Hazlewood    55   103  ...      AUS              No             7

[5 rows x 14 columns]

**GP Swann, a England player,only played 5 years in his career which is the shortes time period in the dataset.**

### **Q6. How many Australian Bowlers are present in this dataset?**

In [502]:
# here len method is used to find the total number of  a fixed country; prior to that specific condition here 'AUS' is imposed in the data set
len(df1[df1['Country'] == 'AUS'])

18

**There are total 18 Australian Bowlers present in the dataset**

### **Q7. Is there any Bangladeshi player present in this dataset?**

In [503]:
print(df1[df1['Country'] == 'BDESH'])
len(df1[df1['Country'] == 'BDESH'])


              Player Mat  Inns  ...  Country  played_for_ICC  years_played
70  Shakib Al Hasan   58    98  ...    BDESH              No            14

[1 rows x 14 columns]


1

**There is only one player from Bangladesh and it's Shakib Al Hasan**

### **Q8. Which player had the lowest economy rate?**

In [504]:
# To find out this, we will take Econ column with ascending equal to true 
df1.sort_values('Econ', ascending = True).head(5)

Player Mat  Inns  ...  Country  played_for_ICC  years_played
32      LR Gibbs   79   148  ...       WI              No            18
47      R Benaud   63   116  ...      AUS              No            12
35  DL Underwood   86   151  ...      ENG              No            16
39       BS Bedi   67   118  ...    INDIA              No            13
68   CV Grimmett   37    67  ...      AUS              No            11

[5 rows x 14 columns]

**LR Gibbs from West Indies is the lowest economy rating player; In one over, he conceded almost 2 runs.**

### **Q9. Which player had the lowest strike rate?**

In [505]:
# again to finout lowest strike rate, we sort out the SR column
df1.sort_values('SR', ascending = True).head(5)


Player Mat  Inns  ...  Country  played_for_ICC  years_played
71      K Rabada   47    86  ...       SA              No             6
7       DW Steyn   93   171  ...       SA              No            15
20  Waqar Younis   87   154  ...      PAK              No            14
19   MD Marshall   81   151  ...       WI              No            13
25     AA Donald   72   129  ...       SA              No            10

[5 rows x 14 columns]

**K Rabada is the lowest strike rate bowler who took the every wickets after 42 balls in his criket career; Another thing is that- First and second lowest strike rate bowler, both are from South Africa.**

### **Q10. Which player had the lowest bowling average?**

In [506]:
# Ave column is sorted to findout the lowedt bowling average 
df1.sort_values('Ave', ascending = True).head(5)


Player  Mat  Inns  ...  Country  played_for_ICC  years_played
19  MD Marshall    81   151  ...       WI              No            13
41     J Garner    58   111  ...       WI              No            10
15  CEL Ambrose    98   179  ...       WI              No            12
33   FS Trueman    67   127  ...      ENG              No            13
4    GD McGrath   124   243  ...      AUS              No            14

[5 rows x 14 columns]

**MD Marshall, a West Indies player, had the lowest bowling average; that means Marshall took one wicket after giving 20.94 runs. And chronologically, three players from West Indies are the lower bowling average than other countries players.**